
# HG changeset patch
# User welpy-cw <h.w.forms@arcor.de>
# Date 1753977045 -7200
# Node ID 1cd9c4860ca15cd28de4d8c90a1522eedf51566f
# Parent  581dc103a22813125a8f1723607d2492e1857497
Bug 1979618 - Fix posting an NNTP message when news server requests auth. r=darktrojan


When a cached NNTP client is reused and the underlying TCP connection must be
re-established, the NNTP server may request authentication when the POST
command is issued. This situation is not handled correctly, as
NNTPClient._currentAction, which holds the action to be repeated after
authentication, is still set to _actionGroup. The message send code then
receives a report of that command's successful completion, giving the
impression that the message had been sent when it actually had not.

This correctly sets _currentAction for posting and has it properly reset
afterwards.

Also reset _messageId and _articleNumber after use, since these may otherwise
prevent the display of an alert dialog in _onData when they should not.

Differential Revision: https://phabricator.services.mozilla.com/D258867

diff --git a/mailnews/news/src/NntpClient.sys.mjs b/mailnews/news/src/NntpClient.sys.mjs
--- a/mailnews/news/src/NntpClient.sys.mjs
+++ b/mailnews/news/src/NntpClient.sys.mjs
@@ -84,16 +84,19 @@ export class NntpClient {
     this.onData = () => {};
     this.onDone = () => {};
 
     this.runningUri = null;
     this.urlListener = null;
     this._msgWindow = null;
     this._newsFolder = null;
     this._nextAction = null;
+    this._currentAction = null;
+    this._messageId = null;
+    this._articleNumber = null;
   }
 
   /**
    * Initiate a connection to the server
    */
   connect() {
     this._done = false;
     if (this._socket?.readyState == "open") {
@@ -535,18 +538,18 @@ export class NntpClient {
   /**
    * Send `POST` request to the server.
    */
   post() {
     const action = () => {
       this._nextAction = this._actionHandlePost;
       this._sendCommand("POST");
     };
+    this._currentAction = action;
     if (this._server.pushAuth && !this._authenticated) {
-      this._currentAction = action;
       this._actionAuthUser();
     } else {
       action();
     }
   }
 
   /**
    * Send `QUIT` request to the server.
diff --git a/mailnews/news/test/unit/test_nntpPostAuthRequired.js b/mailnews/news/test/unit/test_nntpPostAuthRequired.js
new file mode 100644
--- /dev/null
+++ b/mailnews/news/test/unit/test_nntpPostAuthRequired.js
@@ -0,0 +1,50 @@
+// Tests that messages are correctly posted even if the server requests for
+// authentication (Bug 1979618).
+
+var { MailServices } = ChromeUtils.importESModule(
+  "resource:///modules/MailServices.sys.mjs"
+);
+var { PromiseTestUtils } = ChromeUtils.importESModule(
+  "resource://testing-common/mailnews/PromiseTestUtils.sys.mjs"
+);
+
+/* import-globals-from ../../../test/resources/passwordStorage.js */
+load("../../../resources/passwordStorage.js");
+
+add_task(async function test_bug1979618() {
+  // Prepare files for passwords (generated by a script in bug 1018624).
+  await setupForPassword("signons-mailnews1.8.json");
+
+  const daemon = setupNNTPDaemon();
+  const handler = new NNTP_RFC4643_extension(daemon);
+  const server = new nsMailServer(() => handler, daemon);
+  server.start();
+
+  // Send post3.eml to the server.
+  const localServer = setupLocalServer(server.port);
+  const testFile = do_get_file("postings/post3.eml");
+  const urlListener = new PromiseTestUtils.PromiseUrlListener();
+  MailServices.nntp.postMessage(
+    testFile,
+    "test.empty",
+    localServer.key,
+    urlListener,
+    null
+  );
+  await urlListener.promise;
+
+  // Check that the POST command is resend after the server requested
+  // authentication.
+  const transaction = server.playTransaction();
+  do_check_transaction(transaction, [
+    "POST",
+    "AUTHINFO user testnews",
+    "AUTHINFO pass newstest",
+    "POST",
+  ]);
+
+  // Check that the message has been transferred correctly to the server.
+  equal(handler.post, await IOUtils.readUTF8(testFile.path));
+
+  server.stop();
+});
diff --git a/mailnews/news/test/unit/xpcshell.toml b/mailnews/news/test/unit/xpcshell.toml
--- a/mailnews/news/test/unit/xpcshell.toml
+++ b/mailnews/news/test/unit/xpcshell.toml
@@ -35,16 +35,18 @@ disabled = "The server doesn't support r
 ["test_nntpPassword.js"]
 
 ["test_nntpPassword3.js"]
 
 ["test_nntpPasswordFailure.js"]
 
 ["test_nntpPost.js"]
 
+["test_nntpPostAuthRequired.js"]
+
 ["test_nntpProtocols.js"]
 
 ["test_nntpProxy.js"]
 
 ["test_nntpUrl.js"]
 
 ["test_server.js"]
 run-sequentially = true  # Uses fixed NNTP_PORT
diff --git a/mailnews/test/fakeserver/Nntpd.sys.mjs b/mailnews/test/fakeserver/Nntpd.sys.mjs
--- a/mailnews/test/fakeserver/Nntpd.sys.mjs
+++ b/mailnews/test/fakeserver/Nntpd.sys.mjs
@@ -611,9 +611,15 @@ export class NNTP_RFC4643_extension exte
     ) {
       return super.GROUP(args);
     }
     if (this._daemon.groupCredentials != null) {
       this.lastGroupTried = args;
     }
     return "480 Authentication required";
   }
+  POST() {
+    if (this.authenticated) {
+      return super.POST();
+    }
+    return "480 Authentication required";
+  }
 }

